Explain that static Pods never get evicted under node pressure

jayeshmahajan opened the pull request:
<!--
 Hello!

 PLEASE title the FIRST commit appropriately, so that if you squash all
 your commits into one, the combined commit message makes sense.
 For overall help on editing and submitting pull requests, visit:
  https://kubernetes.io/docs/contribute/suggesting-improvements/

 Use the default base branch, “main”, if you're documenting existing
 features in the English localization.

 If you're working on a different localization (not English), see
 https://kubernetes.io/docs/contribute/new-content/overview/#choose-which-git-branch-to-use
 for advice.

 If you're documenting a feature that will be part of a future release, see
 https://kubernetes.io/docs/contribute/new-content/new-features/ for advice.
-->
### Description
Update README: Add priorityClassName usage for static Pods
This PR enhances the README with an example and explanation of how to assign priorityClassName to static Pods.

📌 Includes a note clarifying how kubelet treats static Pods with respect to priority.

✅ Adds sample output of kubectl get priorityclass.

📝 Demonstrates a complete static Pod definition using priorityClassName: xyz-priority.

This addition improves understanding of static Pod scheduling behavior and resource management.
<!--
 Remember to ADD A DESCRIPTION and delete this note before submitting
 your pull request. The description should explain what will change,
 and why.
-->

### Issue
https://github.com/kubernetes/kubernetes/issues/126542

<!--
 If this pull request resolves an open issue, please link the issue in the PR
 description so it will automatically close when the PR is merged.

 See the GitHub documentation for more details and other options:

 https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword
-->

Closes: # https://github.com/kubernetes/kubernetes/issues/126542

jayeshmahajan committed:
Update node-pressure-eviction.md

Static Pods never get evicted when under Node Pressure Issue #126542

jayeshmahajan committed:
Update node-pressure-eviction.md

jayeshmahajan committed:
Update node-pressure-eviction.md

tengqm commented on content/en/docs/concepts/scheduling-eviction/node-pressure-eviction.md:
```diff
@@ -354,8 +354,49 @@ will choose to evict pods of lowest Priority first.
+`priorityClassName` field for that Pod directly. 
+
+**Note:**  
+>The kubelet considers all static pods as critical regardless of their `.spec.priorityClassName`.
```
Are you sure about this?
My take is that static Pods can be evicted as just other Pods. They will never get admitted unless the kubelet is restarted.

tengqm commented on content/en/docs/concepts/scheduling-eviction/node-pressure-eviction.md:
```diff
@@ -354,8 +354,49 @@ will choose to evict pods of lowest Priority first.
+apiVersion: v1
+kind: Pod
+metadata:
+  name: static-web
```
This is an interesting usage scenario ... 
you are running a static Pod as web server?

Static pods are created by the `kubelet`, even when the `kubelet` doesn't know a thing about the API server. In this case, how does the kubelet handle things like priority class (e.g. `priorityClassName`)?

lmktfy replied:
The kubelet creates mirror Pods and responds to (externally managed) static pod manifests.

But yes, I can't see how a static Pod can usefully have a `priorityClassName`.

lmktfy commented:
/retitle Explain that static Pods never get evicted under node pressure

lmktfy commented:
We avoid mentioning issues in commit messages

lmktfy commented on content/en/docs/concepts/scheduling-eviction/node-pressure-eviction.md:
```diff
@@ -354,8 +354,49 @@ will choose to evict pods of lowest Priority first.
-`priorityClassName` field.
+`priorityClassName` field for that Pod directly. 
+
+**Note:**  
```
We have a special short code for note callouts; please use that.

See https://kubernetes.io/docs/contribute/style/style-guide/#note

lmktfy commented on content/en/docs/concepts/scheduling-eviction/node-pressure-eviction.md:
````diff
@@ -354,8 +354,49 @@ will choose to evict pods of lowest Priority first.
+xyz-priority              1000000      false            47s
+```
+
+This is a definition for a static Pod. You would define this directly for the kubelet on a particular nod (and not via kubectl)
````
Typo: nod

jayeshmahajan replied:
fixed

lmktfy commented on content/en/docs/concepts/scheduling-eviction/node-pressure-eviction.md:
```diff
@@ -354,8 +354,49 @@ will choose to evict pods of lowest Priority first.
+**Note:**  
+>The kubelet considers all static pods as critical regardless of their `.spec.priorityClassName`.
+>Static pods also pass kubelet admission even if a node does not have enough resources; the kubelet may
+>instead evict another Pod (potentially even evicting a different Pod at a higher priority, if there is no better match).
```
Can you give an example of the kubelet evicting a higher-priority, non static Pod? It will help folks decide whether to accept the changes in this PR.

jayeshmahajan committed:
Update node-pressure-eviction.md

jayeshmahajan committed:
Update node-pressure-eviction.md

lmktfy commented:
@jayeshmahajan please see https://github.com/kubernetes/website/pull/50569#discussion_r2053061126

jayeshmahajan commented:
> @jayeshmahajan please see [#50569 (comment)](https://github.com/kubernetes/website/pull/50569#discussion_r2053061126)

I have replied there with example.

jayeshmahajan committed:
Update content/en/docs/concepts/scheduling-eviction/node-pressure-eviction.md

Co-authored-by: Tim Bannister <193443691+lmktfy@users.noreply.github.com>

jayeshmahajan committed:
Merge branch 'kubernetes:main' into patch-7

jayeshmahajan committed:
Revise static pod example to StatefulSet format

Updated static pod example to use StatefulSet and clarified priority handling.

jayeshmahajan commented on content/en/docs/concepts/scheduling-eviction/node-pressure-eviction.md:
```diff
@@ -354,8 +354,49 @@ will choose to evict pods of lowest Priority first.
+`priorityClassName` field for that Pod directly. 
+
+{{< note >}}
+>The kubelet considers all static pods as critical regardless of their `.spec.priorityClassName`.
```
ty, fixed

jayeshmahajan committed:
Merge branch 'kubernetes:main' into patch-7

jayeshmahajan committed:
Fix formatting issues in node-pressure-eviction.md

lmktfy commented on content/en/docs/concepts/scheduling-eviction/node-pressure-eviction.md:
````diff
@@ -355,10 +355,70 @@ then the kubelet must choose to evict one of these pods to preserve node stabili
+
+```yaml
+apiVersion: apps/v1
+kind: StatefulSet
````
The text says to expect a Pod. This is a StatefulSet.

I am very dubious about this PR because to me the two things are not easy to mix up.

jayeshmahajan committed:
Its pod not statefulset

Fixing it with correct yaml.

jayeshmahajan committed:
Change Pod kind from StatefulSet and update volumes

jayeshmahajan committed:
added ns from default to redis

Added a command to create a Redis namespace and updated the Pod definition to include the namespace.
